codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,506 @@
-            streamId = localEndpoint.incrementAndGetNextStreamId();
-            try {
-                // Try to create a stream in OPEN state before writing headers, to catch errors on stream creation
-                // early on i.e. max concurrent streams limit reached, stream id exhaustion, etc.
-                localEndpoint.createStream(streamId, false);
-            } catch (Http2Exception e) {
-                promise.setFailure(e);
+            final DefaultHttp2FrameStream stream = (DefaultHttp2FrameStream) headersFrame.stream();
+            final Http2Connection connection = connection();
+            streamId = connection.local().incrementAndGetNextStreamId();
+            if (streamId < 0) {
+                promise.setFailure(new Http2NoMoreStreamIdsException());
                 return;
             }
-            ctx.fireUserEventTriggered(new Http2StreamActiveEvent(streamId, headersFrame));
+            stream.id = streamId;
+
+            numBufferedStreams++;
+            // Ensure that the listener gets executed before any listeners a user might have attached.
+            // TODO(buchgr): Once Http2Stream2 and Http2Stream are merged this is no longer necessary.
+            writePromise = ctx.newPromise();
+            writePromise.addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    numBufferedStreams--;
+
+                    Throwable cause = future.cause();
+                    if (cause == null) {
+
+                        Http2Stream connectionStream  = connection.stream(streamId);
+                        if (connectionStream != null) {
+                            stream.setStreamAndProperty(streamKey, connectionStream);
+                            promise.setSuccess();
+                        } else {
+                            promise.setFailure(new Http2Exception.StreamException(
+                                    streamId, Http2Error.STREAM_CLOSED, "Stream not exists"));
+                        }
+                    } else {
+                        promise.setFailure(cause);
+                    }
+                }
+            });
         }
-        http2Handler.encoder().writeHeaders(http2HandlerCtx, streamId, headersFrame.headers(),
-                                            headersFrame.padding(), headersFrame.isEndStream(), promise);
+        encoder().writeHeaders(ctx, streamId, headersFrame.headers(), headersFrame.padding(),
+                                            headersFrame.isEndStream(), writePromise);
     }
 
     private final class ConnectionListener extends Http2ConnectionAdapter {
+
         @Override
         public void onStreamActive(Http2Stream stream) {
-            if (ctx == null) {
-                // UPGRADE stream is active before handlerAdded().
+            if (isOutboundStream(connection().isServer(), stream.id())) {
```

Comment:
ah got it... let me change it then. Didn't spot this when I looked yesterday, my fault.